### PR TITLE
Fixes Open current TextQuestionDialog using centralized RunCustomDialog

### PR DESCRIPTION
**Pull request by netonjm**
This PR fixes NRE trying to show the TextQuestionDialog because expects a Gtk window but it's getting an NSWindow.  

Just fixed GetParentForModalWindow to also check NSApplication.SharedApplication.ModalWindow and RunCustomDialog now calls CenterWindow.

Probably with this change is also **fixed focus the parent view in natives modal views**, because the parent was not well calculated.

Fixes Bug #802168 - Android app archive and publishing is broken - fails at MessageService.GetPassword.
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/802168

curioso

### CODE REVIEW

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs [hunk 1]
```diff
@@ -594,12 +594,32 @@ public string GetTextResponse (Window parent, string question, string caption, s
 					Question = question,
 					Caption = caption,
 					Value = initialValue,
-					IsPassword = isPassword,
-					TransientFor = parent ?? DesktopService.GetParentForModalWindow ()
+					IsPassword = isPassword
+				};
+
+#if MAC
+				if (parent == null) {
```
Comment (sevoku) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs]: Shouldn't all this happen in [`TextQuestionDialog.RunDefault ()`](https://github.com/mono/monodevelop/blob/ad0f20e38619915304a7bde84013c72ee3c210b3/main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Extensions/TextQuestionDialog.cs#L105) and not here?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs [hunk 2]
```diff
@@ -594,12 +594,32 @@ public string GetTextResponse (Window parent, string question, string caption, s
...
+					parent = NSApplication.SharedApplication.ModalWindow;
+				}
+#endif
+				if (parent == null) {
+					parent = DesktopService.GetParentForModalWindow ();
+				}
+
+				var customDialog = dialog.Create ();
+				customDialog.Shown += (sender, e) => {
+					CenterWindow (customDialog, parent);
```
Comment (sevoku) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs]: `RunCustomDialog` should do that, if it doesn't then it's a bug there. It shouldn't be necessary to call it here again. See also #7264

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Extensions/TextQuestionDialog.cs [hunk 3]
```diff
@@ -62,40 +62,53 @@ public class TextQuestionDialog : PlatformDialog<TextQuestionDialogData>
 			get { return data.IsPassword; }
 			set { data.IsPassword = value; }
 		}
-		
+
+		public Gtk.Dialog Create ()
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Extensions/TextQuestionDialog.cs]: Should this be public API?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs (hunk 1 above)
Comment (netonjm) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs]: Yes! agree!

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs (hunk 2 above)
Comment (netonjm) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs]: 👍

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Extensions/TextQuestionDialog.cs [hunk 4]
```diff
@@ -62,40 +62,53 @@ public class TextQuestionDialog : PlatformDialog<TextQuestionDialogData>
...
-
-				md.DefaultResponse = ResponseType.Cancel;
-				
-				md.Child.ShowAll ();
-				
+				md = Create ();
 				var response = (ResponseType) MonoDevelop.Ide.MessageService.RunCustomDialog (md, TransientFor);
 				if (response == ResponseType.Ok) {
-					Value = responseEntry.Text;
 					return true;
```
Comment (rodrmoya) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Extensions/TextQuestionDialog.cs]: why was this moved to the `Response` event handler?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs (hunk 2 above)
Comment (rodrmoya) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs]: Yes please, just run everything via `RunCustomDialog`, it already has code to show, parent and centre the window correctly. In fact, this code should just create the dialog and call its `RunDefault`. No need to create the dialog here, right?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Extensions/TextQuestionDialog.cs (hunk 3 above)
Comment (sevoku) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Extensions/TextQuestionDialog.cs]: no, it shouldn't

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs (hunk 2 above)
Comment (netonjm) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs]: Fixed

**Comment by netonjm** [comment]
Hey guys I changed completely the approach could you take a look? @rodrmoya  @sevoku  @Therzok

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs [hunk 5]
```diff
@@ -348,9 +348,14 @@ public static int RunCustomDialog (Dialog dlg, Window parent)
...
 				// If there is a native NSWindow model window running, we need
 				// to show the new dialog over that window.
-				if (NSApplication.SharedApplication.ModalWindow != null)
-					dialog.Shown += HandleShown;
-				else
+				if (NSApplication.SharedApplication.ModalWindow != null) {
+					EventHandler shownHandler = null;
+					shownHandler = (s,e) => {
+						ShowCustomModalDialog (dialog, parent);
+						dialog.Shown -= shownHandler;
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs]: What value of shownHandler is captured here? It might not unsubscribe the event?

Comment (netonjm) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs]: that's is a common pattern, no?

shownHandler is suscribed on shown:

https://github.com/mono/monodevelop/pull/7265/files/0e847640b8f5b6e33711ad6db6f31c678f7ebc3b#diff-e33cd0734d92a14cc69170d8404ffca3R353

and unsubscribed when finished:

https://github.com/mono/monodevelop/blob/0e847640b8f5b6e33711ad6db6f31c678f7ebc3b/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs#L357 

it's not correct?

Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/MessageService.cs]: Just tested a small app, it's working fine!

**Comment by slluis** [comment]
@monojenkins backport release-8.0-integration
